src/System.Memory/ref/System.Memory.cs

```diff
@@ -60,6 +60,11 @@ public static partial class MemoryExtensions
         public static bool Overlaps<T>(this System.ReadOnlySpan<T> first, System.ReadOnlySpan<T> second, out int elementOffset) { throw null; }
         public static bool Overlaps<T>(this System.Span<T> first, System.ReadOnlySpan<T> second) { throw null; }
         public static bool Overlaps<T>(this System.Span<T> first, System.ReadOnlySpan<T> second, out int elementOffset) { throw null; }
+        public static void PadLeft(this System.ReadOnlySpan<char> source, int totalWidth, System.Span<char> destination) { }
+        public static void PadLeft(this System.ReadOnlySpan<char> source, int totalWidth, System.Span<char> destination, char paddingChar) { }
```

Comment:
We should answer this question by trying this API out on several real-world examples.

Reply:
Some examples we can look at [WIP]:
https://github.com/PowerShell/PowerShell/blob/4bc52d2358222084738157a08907fac32d13bd3a/src/Microsoft.PowerShell.Commands.Utility/commands/utility/UtilityCommon.cs#L168

https://github.com/ShareX/ShareX/search?utf8=%E2%9C%93&q=PadLeft&type=

This one is interesting [WIP]:
https://github.com/JeffreySu/WeiXinMPSDK/blob/5d62c88344f793d9da471670e31d989f5a3a542a/src/Senparc.Weixin/Senparc.Weixin/Helpers/EncryptHelper.cs#L268
```C#
Byte[] bKey = new Byte[32];
Array.Copy(Encoding.UTF8.GetBytes(Key.PadRight(bKey.Length)), bKey, bKey.Length);
```

Can be written as:
```C#
Byte[] bKey = new Byte[32];
Span<char> result = new char[bKey.Length];
...(...(Key.AsReadOnlySpan().PadRight(result))...);
```
Atm, it would be:
```C#
Byte[] bKey = new Byte[32];
Span<char> result = new char[bKey.Length];
...(...(Key.AsReadOnlySpan().PadRight(result.Length, result))...);
```

Reply:
> UtilityCommon.cs#L168

This looks like overengineered code. It can be written in much simpler way using `X8` formatting string.

> Byte[] bKey = new Byte[32];
> Span<char> result = new char[bKey.Length];
> ...(...(Key.AsReadOnlySpan().PadRight(result))...);

This has same number, or one more extra, allocation compared to original code. Switching to Span should naturally minimize or eliminate allocations. Otherwise, it is just an unnecessary churn.

Reply:
> This has same number, or one more extra, allocation compared to original code. 

Fixed.

```C#
Byte[] bKey = new Byte[32];
Span<byte> bKeySpan = bKey;
Span<char> bKeyChar = bKeySpan.NonPortableCast<byte, char>();
Key.AsReadOnlySpan().PadRight(bKeyChar.Length, bKeyChar);
Encodings.Utf16.ToUtf8Length(bKeySpan, out int bytesNeeded);
if (bytesNeeded > bKey.Length) throw new Exception();
OperationStatus result = Encodings.Utf16.ToUtf8(bKeySpan, bKeySpan, out int consumed, out int written);
if (result != OperationStatus.Done) throw new Exception();
```

>  Switching to Span should naturally minimize or eliminate allocations.



If we include our [encoding](https://github.com/dotnet/corefxlab/blob/master/src/System.Text.Primitives/System/Text/Encoders/Utf16.cs#L52) APIs, the span version definitely allocates less.

Reply:
So you have replaced one easy to understand line with 8 lines of hard to understand code, that also uses non-portable APIs as bonus. This does not look like an improvement to me. 

All this is in a crypto-stream initialization that allocates a ton of other stuff, so a few extra bytes saved is not going to make a difference.

Reply:
That's fair. My intent was try to answer this question:

> I don't think we should have totalWidth in Span APIs. The width of the resulting string should be the size of the destination span.

From the uses of PadLeft/PadRight, the totalWidth argument seems redundant. I rarely see the totalWidth < source.Length case.